kin-devplatform-sdk/src/main/java/kin/devplatform/KinEcosystemInitiator.java

```diff
@@ -0,0 +1,265 @@
+
+			DeviceUtils.init(context);
+
+			setAppID();
+
+			eventLogger.send(KinSdkInitiated.create());
+			isInitialized = true;
+		}
+	}
+
+	private void setAppID() {
+		ObservableData<String> observableData = AuthRepository.getInstance().getAppID();
+		String appID = observableData.getValue();
+		if (appID == null) {
+			observableData.addObserver(new Observer<String>() {
+				@Override
+				public void onChanged(String appID) {
+					if (!TextUtils.isEmpty(appID)) {
+						BlockchainSourceImpl.getInstance().setAppID(appID);
+						AuthRepository.getInstance().getAppID().removeObserver(this);
+					}
+				}
+			});
+		}
+		BlockchainSourceImpl.getInstance().setAppID(appID);
+	}
+
+	private void login(@NonNull SignInData signInData, final KinCallback<Void> loginCallback) {
+		setAuthRepositoryData(signInData);
+		performLogin(loginCallback);
+	}
+
+	private void setAuthRepositoryData(@NonNull SignInData signInData) {
+		String publicAddress = BlockchainSourceImpl.getInstance().getPublicAddress();
+		String deviceID = AuthRepository.getInstance().getDeviceID();
+		signInData.setDeviceId(deviceID != null ? deviceID : UUID.randomUUID().toString());
+		signInData.setWalletAddress(publicAddress);
+		AuthRepository.getInstance().setSignInData(signInData);
+	}
+
+	private void performLogin(final KinCallback<Void> loginCallback) {
+		AuthRepository.getInstance().getAuthToken(new KinCallback<AuthToken>() {
+			@Override
+			public void onResponse(AuthToken response) {
+				final AccountManager accountManager = AccountManagerImpl.getInstance();
+				if (accountManager.isAccountCreated()) {
+					activateAccount(loginCallback);
+				} else {
+					handleAccountNotCreatedState(accountManager, loginCallback);
+				}
+			}
+
+			@Override
+			public void onFailure(final KinEcosystemException exception) {
+				fireStartError(exception, loginCallback);
+			}
+		});
+	}
+
+	private void handleAccountNotCreatedState(final AccountManager accountManager,
+		final KinCallback<Void> loginCallback) {
+		final Handler handler = new Handler(Looper.getMainLooper());
+		final Observer<Integer> accountStateObserver = new Observer<Integer>() {
+			@Override
+			public void onChanged(@AccountState Integer value) {
+				if (value == AccountManager.ERROR) {
+					handler.removeCallbacksAndMessages(null);
+					accountManager.removeAccountStateObserver(this);
+					fireStartError(accountManager.getError(), loginCallback);
+				} else if (value == AccountManager.CREATION_COMPLETED) {
+					handler.removeCallbacksAndMessages(null);
+					accountManager.removeAccountStateObserver(this);
+					activateAccount(loginCallback);
+				}
+			}
+		};
+
+		handler.postDelayed(new Runnable() {
+			@Override
+			public void run() {
+				accountManager.removeAccountStateObserver(accountStateObserver);
+				fireStartError(ErrorUtil.createCreateAccountTimeoutException(), loginCallback);
+			}
+		}, ACCOUNT_CREATION_TIME_OUT_MILLIS);
```

Comment:
Without getting into too much detail on the "why" but just from a look a this handler, specifically because it is post delay on some time that probably other method also count on then it is seems a bit "fishy" and maybe not to synchronized, if you think it is fine in this context then it is ok. 
Maybe you can somehow use callback instead of counting on timers